### PR TITLE
tweak `when` for kill active terminal editor command

### LINKED ISSUE #188145: Ctrl+F4 closes Terminal in Editor area only if I modify default context settings
Type: <b>Bug</b>

I had to modify:
```
{
  "key": "ctrl+f4",
  "command": "workbench.action.terminal.killEditor",
  "when": "terminalEditorFocus && terminalFocus && terminalHasBeenCreated && resourceScheme == 'vscode-terminal' || terminalEditorFocus && terminalFocus && terminalProcessSupported && resourceScheme == 'vscode-terminal'"
}
```
to
``` 
{
  "key": "ctrl+f4",
  "command": "workbench.action.terminal.killEditor",
  "when": "terminalEditorFocus && terminalFocus && terminalHasBeenCreated || terminalEditorFocus && terminalFocus && terminalProcessSupported"
}
```
for the Ctrl+F4 to actually close a terminal that was dragged to editor area. Same for Ctrl+W.

image

VS Code version: Code 1.80.1 (74f6148eb9ea00507ec113ec51c489d6ffb4b771, 2023-07-12T17:22:07.651Z)
OS version: Windows_NT x64 10.0.22000
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz (8 x 2803)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.65GB (2.50GB free)|
|Process Argv|--crash-reporter-id 2ef976af-0ebf-4a25-859c-0d6742e63bdd|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (77)</summary>

Extension|Author (truncated)|Version
---|---|---
better-comments|aar|3.0.2
capnproto-syntax|abr|0.1.0
nugetpackagemanagergui|ali|2.0.6
QML|bbe|1.0.0
doxdocgen|csc|1.4.0
auto-header|Dan|1.0.1
githistory|don|0.6.20
python-environment-manager|don|1.0.4
gitlens|eam|14.1.0
chatgpt-gpt4-gpt3-vscode|Eas|1.0.2
code-runner|for|0.12.0
visual-nuget|ful|0.2.1
chatgpt-vscode|gen|0.0.8
remotehub|Git|0.60.0
vscode-github-actions|git|0.25.8
vscode-pull-request-github|Git|0.68.1
go|gol|0.39.1
todo-tree|Gru|0.0.226
language-cython|guy|0.0.7
vscode-test-explorer|hbe|2.21.1
output-colorizer|IBM|0.1.2
pack-vsix|igo|0.0.91
vscode-edit-csv|jan|0.7.6
c-cpp-flylint|jbe|1.13.0
better-cpp-syntax|jef|1.17.2
cmake-language-support-vscode|jos|0.0.9
hex-fmt|ker|1.0.0
octave|lea|0.1.1
dbc|lha|2.0.0
cortex-debug|mar|1.12.0
debug-tracker-vscode|mcu|0.0.15
memory-view|mcu|0.0.23
peripheral-viewer|mcu|1.4.6
rtos-views|mcu|0.0.7
rainbow-csv|mec|3.7.0
git-graph|mhu|1.30.0
csharp|ms-|1.26.0
vscode-dotnet-runtime|ms-|1.6.0
jupyter|ms-|2023.6.1101941928
jupyter-keymap|ms-|1.1.2
jupyter-renderers|ms-|1.0.17
vscode-jupyter-cell-tags|ms-|0.1.8
vscode-jupyter-slideshow|ms-|0.1.5
remote-containers|ms-|0.299.0
remote-ssh|ms-|0.102.0
remote-ssh-edit|ms-|0.86.0
remote-wsl|ms-|0.80.2
vscode-remote-extensionpack|ms-|0.24.0
azure-repos|ms-|0.36.0
cmake-tools|ms-|1.14.34
cpptools|ms-|1.16.3
cpptools-extension-pack|ms-|1.3.0
hexeditor|ms-|1.9.11
makefile-tools|ms-|0.7.0
remote-explorer|ms-|0.4.1
remote-repositories|ms-|0.38.1
test-adapter-converter|ms-|0.1.8
vscode-typescript-next|ms-|5.2.20230717
clang-tidy|not|0.5.1
vsc-octave-debugger|pau|0.5.11
platformio-ide|pla|3.3.0
pokowaka-iwyu|pok|1.0.3
quantum-devkit-vscode|qua|0.28.277227
vscode-yaml|red|1.13.0
c-cpp-definition-generator|rei|0.0.2
rust-analyzer|rus|0.3.1591
codify|sma|1.0.45
sonarlint-vscode|Son|3.19.2
cody-ai|sou|0.6.0
remote-x11-ssh|spa|1.5.0
rewrap|stk|1.16.3
octave|toa|0.0.3
gnu-mapfiles|tro|1.1.0
cmake|twx|0.0.17
keyoti-changeallendoflinesequence|vs-|0.0.3
code-d|web|0.23.2
wokwi-vscode|Wok|1.3.5

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30410667
vsclangdc:30486549
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
vscccc:30610679
3biah626:30602489
pyind779:30671433
89544117:30613380
vscrpc:30673769
pythonsymbol12:30671437
2i9eh265:30646982
showlangstatbar:30737416
vsctsb:30748421
pythonms35:30701012
03d35959:30757346
pythonfmttext:30731395
pythoncmvfstrcf:30756944
fixshowwlkth:30771522
hideindicator:30785051
pythongtdpath:30769146
i26e3531:30780429
pythonnosm12tcf:30779713
pythonidxpt:30784022
pythonnocebcf:30776496
e537b577:30786199
dsvsc013:30789517

```

</details>

### PR DESCRIPTION
**Pull request by meganrogge**
fix #188145

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminalActions.ts [hunk 1]
```diff
@@ -1312,7 +1311,7 @@ export function registerTerminalActions() {
 			primary: KeyMod.CtrlCmd | KeyCode.KeyW,
 			win: { primary: KeyMod.CtrlCmd | KeyCode.F4, secondary: [KeyMod.CtrlCmd | KeyCode.KeyW] },
 			weight: KeybindingWeight.WorkbenchContrib,
-			when: ContextKeyExpr.and(TerminalContextKeys.focus, ResourceContextKey.Scheme.isEqualTo(Schemas.vscodeTerminal), TerminalContextKeys.editorFocus)
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: Do you know why it doesn't match that? It's meant to as it's a terminal editor.

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: when I focus run `inspect context keys` and focus a terminal, here are the results. See how `resourceScheme` is `file`

Screenshot 2023-11-15 at 9 35 34 AM

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: The question I'm curious about then is why this URI isn't coming through:

https://github.com/microsoft/vscode/blob/e2670a457fbba846c8adebc70bcaec674c949866/src/vs/workbench/contrib/terminal/browser/terminalEditorInput.ts#L122-L123

Provided it's actually still vscode-terminal like it used to be.

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: logging here and looks like it gets set correctly 🤔 so not sure why it doesn't work

https://github.com/hamirmahal/vscode/blob/0edd8b42dbbcbb682e59b9d8425466e37daa6420/src/vs/workbench/common/contextkeys.ts#L245

https://github.com/microsoft/vscode/assets/29464607/60d9be78-7896-45c8-bd19-682331d3c298

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: also worked when I created a terminal in the editor area directly

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1315]: Ah ok, this works just fine on mac. It does not work on windows, looking into it.

**Comment by meganrogge** [comment]
On Windows:

```
023-11-15 09:51:13.638 [info] [KeybindingService]: \ From 3 keybinding entries, matched workbench.action.closeActiveEditor, when: no when condition, source: built-in.
2023-11-15 09:51:13.753 [info] [KeybindingService]: / Soft dispatching keyboard event
2023-11-15 09:51:13.753 [info] [KeybindingService]: | Resolving ctrl+F4
2023-11-15 09:51:13.753 [info] [KeybindingService]: \ From 3 keybinding entries, matched workbench.action.closeActiveEditor, when: no when condition, source: built-in.
2023-11-15 09:51:13.910 [info] [KeybindingService]: / Soft dispatching keyboard event
2023-11-15 09:51:13.911 [info] [KeybindingService]: \ Keyboard event cannot be dispatched
2023-11-15 09:51:13.911 [info] [KeybindingService]: + Ignoring single modifier ctrl due to it being pressed together with other keys.
```

**Review by Tyriar — APPROVED** [review]
It's probably sufficient, it would be good to understand the problem as we use this in a bunch of places:

image

It looks like the when clause in the editor context menu is working fine 🤔

**Comment by meganrogge** [comment]
Yeah @Tyriar everything else was working and I spent awhile investigating so idk 🤷🏼‍♀️

**Pull request merged**